### PR TITLE
Camel 10327, New Apache drill component

### PR DESCRIPTION
**Pull request by Fabryprog**
https://issues.apache.org/jira/browse/CAMEL-10327

### CODE REVIEW

**Comment by Fabryprog** [comment]
in pom.xml i put 2.19.0-SNAPSHOT version but i want merge it into 2.18.1.

what do i do to fix it?

**Comment by oscerd** [comment]
This is a new component. It can go only on 2.19.x release.

We don't add new components on minor releases like 2.18.1.

**Comment by Fabryprog** [comment]
@oscerd i must push also "components-starter/camel-drill-starter" auto-generated directory?

**Comment by oscerd** [comment]
Yes :-) It is generated during build for Spring-boot Camel Drill starter. Thanks!

File: components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+            url += "/" + clusterId;
+        }
+
+        return url;
+    }
+
+    @SuppressWarnings("unchecked")
+    public List<?> queryForList(ResultSet rs) throws SQLException {
+        ColumnMapRowMapper rowMapper = new ColumnMapRowMapper();
+        RowMapperResultSetExtractor<Map<String, Object>> mapper = new RowMapperResultSetExtractor<Map<String, Object>>(rowMapper);
```
Comment (Fabryprog) [inline components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java]: @lburgazzoli this code is an extract of SqlProducer.java code.

Yes, it is also nice! After this PR i could create another JIRA to improve component.

File: components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    private String host;
+    @UriParam(description = "Port number")
+    @Metadata(required = "false", defaultValue = "2181")
+    private Integer port = 2181;
+    @UriParam(description = "Drill directory", defaultValue = "")
+    private String directory = "";
+    @UriParam(defaultValue = "")
+    private String clusterId = "";
+    @UriParam(defaultValue = "zk")
+    private String mode = "zk";
```
Comment (lburgazzoli) [inline components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java]: I think you can use the enum DrillConnectionMode here, and list possible values in UriParam

File: components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java (hunk 1 above)
Comment (lburgazzoli) [inline components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java]: This does not mean that it is the best option ;)
We should try to write GC-friendly code

File: components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java (hunk 2 above)
Comment (Fabryprog) [inline components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java]: You were right now is better!!!

**Comment by Fabryprog** [comment]
@oscerd it is tested it into my project and working fine (producer is too simple!). it is possible that in the future add other features

**Comment by oscerd** [comment]
@Fabryprog I'll review and merge on Monday. Thanks a lot! :-)

File: components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java (hunk 2 above)
Comment (davsclaus) [inline components/camel-drill/src/main/java/org/apache/camel/component/drill/DrillEndpoint.java]: I think a default value of empty string is better to be null.

**Comment by oscerd** [comment]
You still need to add the related Karaf feature, like camel-asterisk for example.

https://github.com/apache/camel/blob/master/platforms/karaf/features/src/main/resources/features.xml#L120-L124

With the related test, like

https://github.com/apache/camel/blob/master/tests/camel-itest-karaf/src/test/java/org/apache/camel/itest/karaf/CamelAsteriskTest.java

**Comment by Fabryprog** [comment]
@oscerd i put feature like example.

How do test it?

With **mvn clean test** i have follow error:

`2016-10-17 09:37:57,012 | INFO  | pool-7-thread-1  | FeaturesServiceImpl              | 8 - org.apache.karaf.features.core - 4.0.7 | Adding features: instance/[4.0.7,4.0.7], package/[4.0.7,4.0.7], log/[4.0.7,4.0.7], ssh/[4.0.7,4.0.7], aries-blueprint/[4.0.7,4.0.7], framework/[4.0.7,4.0.7], system/[4.0.7,4.0.7], feature/[4.0.7,4.0.7], shell/[4.0.7,4.0.7], management/[4.0.7,4.0.7], service/[4.0.7,4.0.7], jaas/[4.0.7,4.0.7], shell-compat/[4.0.7,4.0.7], deployer/[4.0.7,4.0.7], diagnostic/[4.0.7,4.0.7], wrap/[0,0.0.0], bundle/[4.0.7,4.0.7], config/[4.0.7,4.0.7], kar/[4.0.7,4.0.7], camel/[2.19.0.SNAPSHOT,2.19.0.SNAPSHOT], exam/[4.9.1,4.9.1], test-dependencies/[0,0.0.0]
2016-10-17 09:37:59,035 | ERROR | pool-7-thread-1  | BootFeaturesInstaller            | 8 - org.apache.karaf.features.core - 4.0.7 | Error installing boot features
org.osgi.service.resolver.ResolutionException: Unable to resolve root: missing requirement [root] osgi.identity; osgi.identity=camel; type=karaf.feature; version="[2.19.0.SNAPSHOT,2.19.0.SNAPSHOT]"; filter:="(&(osgi.identity=camel)(type=karaf.feature)(version>=2.19.0.SNAPSHOT)(version<=2.19.0.SNAPSHOT))" [caused by: Unable to resolve camel/2.19.0.SNAPSHOT: missing requirement [camel/2.19.0.SNAPSHOT] osgi.identity; osgi.identity=camel-core; type=karaf.feature; version="[2.19.0.SNAPSHOT,2.19.0.SNAPSHOT]" [caused by: Unable to resolve camel-core/2.19.0.SNAPSHOT: missing requirement [camel-core/2.19.0.SNAPSHOT] osgi.identity; osgi.identity=org.apache.camel.camel-commands-core; type=osgi.bundle; version="[2.19.0.SNAPSHOT,2.19.0.SNAPSHOT]"; resolution:=mandatory [caused by: Unable to resolve org.apache.camel.camel-commands-core/2.19.0.SNAPSHOT: missing requirement [org.apache.camel.camel-commands-core/2.19.0.SNAPSHOT] osgi.wiring.package; filter:="(&(osgi.wiring.package=org.apache.camel.catalog)(version>=2.19.0)(!(version>=2.20.0)))"]]]
	at org.apache.felix.resolver.ResolutionError.toException(ResolutionError.java:42)[8:org.apache.karaf.features.core:4.0.7]
	at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:235)[8:org.apache.karaf.features.core:4.0.7]
`

**Comment by oscerd** [comment]
There is no apachedrill bundle in 

http://repo2.maven.org/maven2/org/apache/servicemix/bundles/

You'll need to wrap the drill bundle

So remove.

```xml
<bundle dependency='true'>mvn:org.apache.servicemix.bundles/org.apache.servicemix.bundles.apachedrill/${apache-drill-bundle-version}</bundle>
```

and add

```xml
<bundle dependency='true'>wrap:mvn:org.apache.drill.exec/drill-jdbc-all/${apache-drill-bundle-version}</bundle>
```

Then build the project inside platforms/karaf/feature (mvn clean install) and in tests/camel-itest-karaf run 

```bash
mvn clean test -Dtest=CamelDrillTest
```

**Comment by oscerd** [comment]
To be sure everything is aligned run a full build of camel with mvn clean install -DskipTests

**Comment by Fabryprog** [comment]
@oscerd i fix it... sorry but it is my first time with osgi bundles!

**Comment by oscerd** [comment]
@Fabryprog , no problem. Good for the next time :-)

In Camel we take care of OSGi support too.

**Comment by oscerd** [comment]
Thanks. The PR has been merged. Can you close this?

**Pull request closed without merging**